Add multi-modal search blog post (#3857)
opensearch-project/project-website: _posts/2025-07-07-building-a-multi-modal-search-engine-in-opensearch.md

---
layout: post
title:  "Building a multimodal search engine in OpenSearch"
authors:
 - mingshl
 - kolchfa

> date: 2025-07-07

Reviewer: Publish date: 2025-07-29

categories:
 - technical-post
meta_keywords: "multimodal search, opensearch, semantic search, vector search, image search"

> meta_description: "Learn how to build a multimodal search engine in OpenSearch using the ML inference ingest and search request processors to search across text and images."

Reviewer: Update meta: 
_meta_keywords:_ multimodal search, multimodality, semantic search, vector embeddings, machine learning, multimodal search in OpenSearch, deep neural networks, multimodal embedding, Titan Multimodal Embeddings, ML inference processor, k-NN vector

_meta_description:_ Learn how to build a powerful multimodal search engine in OpenSearch that processes text, images, and other data types simultaneously using ML inference processors and Amazon Titan models.

```json
PUT _ingest/pipeline/ml_inference_pipeline_multi_modal
{
  "processors": [
    {
      "ml_inference": {
        "tag": "ml_inference",
        "description": "This processor is going to run ml inference during ingest to create multimodal embeddings",
        "model_id": "ncdJJZUB7judm8f4HN3P", // Replace with your model_id
"input_map": [
          {
            "inputText": "name",  // Map document's 'name' field to model's 'inputText'
            "inputImage":"image"   // Map document's 'image' field to model's 'inputImage'
          }
        ],
        "output_map": [
          {
            "multimodal_embedding": "embedding" // Map model's 'embedding' output to document's 'multimodal_embedding' field
          }
        ],
        "ignore_missing":true,
        "ignore_failure": true
      }
    }
  ]
}
```

This ingest pipeline automatically generates a `multimodal_embedding` vector for documents that have either a `name` field (for text), an `image` field (for Base64-encoded images), or both. If neither is present, the processor is skipped.

### 3. Create a vector index

> Now, create a vector index and define a `knn_vector` field for your multimodal embeddings. This field will store the vectors generated by the ingest pipeline:

Reviewer 2: ```suggestion
Now create a vector index and define a `knn_vector` field for your multimodal embeddings. This field will store the vectors generated by the ingest pipeline:
```

            "inputImage": "ext.ml_inference.image" // Map query's 'ext.ml_inference.image' to model's 'inputImage'
          }
        ],
        "output_map": [
          {
            "multimodal_embedding": "embedding" // Map model's 'embedding' output to a variable for the k-NN query
          }
        ],
        "model_config":{},
        "ignore_missing":false,
        "ignore_failure": false
      }
    }
  ]
}
```

Now you can perform multimodal searches!

#### Search using text

The following query takes the text "place where recreational activities are done, picnics happen there", generates an embedding, and then finds documents containing similar multimodal embeddings:

```json
GET test-index-area/_search?search_pipeline=multimodal_semantic_search_pipeline
{
  "query": {
    "match_all": {}
  },
  "ext": {
    "ml_inference": {
      "text": "place where recreational activities are done, picnics happen there"
    }
  }
}
```

This query returns documents, in which the `multimodal_embedding` is closest to the embedding of your text query, effectively finding relevant documents whether they contain matching text or visually similar content.

## Conclusion

> Multimodal search enables more flexible ways to retrieve information, especially when working with diverse types of data, from e-commerce products to digital assets. Using OpenSearch, you can build highly intelligent and flexible search experiences, useful in a range of applications—from searching for products using images to organizing digital assets using text-based queries. 

Reviewer 2: ```suggestion
Multimodal search enables more flexible ways to retrieve information, especially when working with diverse types of data. Using OpenSearch, you can build highly intelligent and flexible search experiences that are useful in a range of applications—from searching for products using images to organizing digital assets using text-based queries. 
```
